keymaster/ta/asn1.c

```diff
@@ -15,9 +15,52 @@
  * limitations under the License.
  */
 
+#include <mbedtls/x509.h>
+
 #include "asn1.h"
 #include "attestation.h"
 
+
+static unsigned int get_key_usage(keymaster_key_param_t *param)
+{
+	unsigned int key_usage = 0;
+
+	switch (param->tag) {
+	case KM_TAG_PURPOSE:
+		DMSG("key purpose 0x%08X",param->key_param.enumerated);
+		if ((param->key_param.enumerated == KM_PURPOSE_VERIFY) ||
+			(param->key_param.enumerated == KM_PURPOSE_SIGN))
+		{
+			key_usage = MBEDTLS_X509_KU_DIGITAL_SIGNATURE;
+		}
+		if ((param->key_param.enumerated == KM_PURPOSE_ENCRYPT) ||
+			(param->key_param.enumerated == KM_PURPOSE_DECRYPT))
+		{
+			key_usage |= MBEDTLS_X509_KU_KEY_ENCIPHERMENT | MBEDTLS_X509_KU_DATA_ENCIPHERMENT;
+		}
+		break;
+	default:
+		DMSG("Unused parameter tag %x", param->tag);
+		break;
+	}
+
+	return key_usage;
+}
+
+static unsigned int add_key_usage(keymaster_key_characteristics_t *key_chr)
+{
+	unsigned int key_usage = 0;
+
+	for (size_t i = 0; i < key_chr->sw_enforced.length; i++) {
```

Comment:
Do we need to parse sw_enforced params for purpose tag?
From what I can see - there is no scenario when this tag can get there.

Reply:
This function is mainly inspired from the equivalent one in Trusty:
add_key_usage_extension at line 199 in system/keymaster/km_openssl/attestation_utils.cpp used in generate_attestation.

Reply:
yeah, I know that sometimes Google doc is not enought, and its easier to see how it is implemented. But we can't just copy from there - default implementation requires sw_enforced domain is widely used, because it is a software implementation.
We have a list of tags that are sw_enforced - and purpose is not one of them, so it should never happen that purpose value is in sf_enforced.